service/worker/deployment/deployment_util.go

```diff
@@ -0,0 +1,226 @@
+	d.deployment.BuildId = d.EscapeChar(d.deployment.BuildId)
+
+	// validate params which are used for building workflowID's
+	err := d.ValidateDeploymentWfParams("DeploymentName", d.deployment.DeploymentName, maxIDLengthLimit)
+	if err != nil {
+		return err
+	}
+	err = d.ValidateDeploymentWfParams("BuildID", d.deployment.BuildId, maxIDLengthLimit)
+	if err != nil {
+		return err
+	}
+
+	deploymentWorkflowID := d.GenerateDeploymentWorkflowID()
+	workflowInputPayloads, err := d.GenerateStartWorkflowPayload()
+	if err != nil {
+		return err
+	}
+	updatePayload, err := d.GenerateUpdateDeploymentPayload(taskQueueName, taskQueueType, pollTimestamp)
+	if err != nil {
+		return err
+	}
+
+	// Start workflow execution, if it hasn't already
+	startReq := &workflowservice.StartWorkflowExecutionRequest{
+		Namespace:                d.namespaceEntry.Name().String(),
+		WorkflowId:               deploymentWorkflowID,
+		WorkflowType:             &commonpb.WorkflowType{Name: DeploymentWorkflowType},
+		TaskQueue:                &taskqueuepb.TaskQueue{Name: primitives.PerNSWorkerTaskQueue},
+		Input:                    workflowInputPayloads,
+		WorkflowIdReusePolicy:    enumspb.WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE,
+		WorkflowIdConflictPolicy: enumspb.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING,
+		SearchAttributes:         sa,
+	}
+
+	updateReq := &workflowservice.UpdateWorkflowExecutionRequest{
+		Namespace: d.namespaceEntry.Name().String(),
+		WorkflowExecution: &commonpb.WorkflowExecution{
+			WorkflowId: deploymentWorkflowID,
+		},
+		Request: &updatepb.Request{
+			Input: &updatepb.Input{Name: RegisterWorkerInDeployment, Args: updatePayload},
+		},
+		WaitPolicy: &updatepb.WaitPolicy{LifecycleStage: enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED},
+	}
+
+	// This is an atomic operation; if one operation fails, both will.
+	_, err = d.historyClient.ExecuteMultiOperation(ctx, &historyservice.ExecuteMultiOperationRequest{
+		NamespaceId: d.namespaceEntry.ID().String(),
+		WorkflowId:  deploymentWorkflowID,
+		Operations: []*historyservice.ExecuteMultiOperationRequest_Operation{
+			{
+				Operation: &historyservice.ExecuteMultiOperationRequest_Operation_StartWorkflow{
+					StartWorkflow: &historyservice.StartWorkflowExecutionRequest{
+						NamespaceId:  d.namespaceEntry.ID().String(),
+						StartRequest: startReq,
+					},
+				},
+			},
+			{
+				Operation: &historyservice.ExecuteMultiOperationRequest_Operation_UpdateWorkflow{
+					UpdateWorkflow: &historyservice.UpdateWorkflowExecutionRequest{
+						NamespaceId: d.namespaceEntry.ID().String(),
+						Request:     updateReq,
+					},
+				},
+			},
+		},
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GenerateDeploymentWorkflowID is a helper that generates a system accepted
+// workflowID which are used in our deployment workflows
+func (d *DeploymentWorkflowClient) GenerateDeploymentWorkflowID() string {
```

Comment:
Better to make the methods that are not part of the public interface private. Same for other methods.

```suggestion
func (d *DeploymentWorkflowClient) generateDeploymentWorkflowID() string {
```